opencog/rule-engine/forwardchainer/ForwardChainer.cc

```diff
@@ -144,19 +147,50 @@ void ForwardChainer::do_chain(ForwardChainerCallBack& fcb,
      init_sources = _as.get_outgoing(hsource);
     else
         init_sources.push_back(hsource);
+
     _fcmem.update_potential_sources(init_sources);
 
-    _fcmem.set_source(fcb.choose_next_source(_fcmem)); //set initial source
     auto max_iter = _configReader.get_maximum_iterations();
 
     while (_iteration < max_iter /*OR other termination criteria*/) {
+
         _log->info("Iteration %d", _iteration);
 
-        do_step(fcb);
+        if(rule_weight.empty())
+        {
+            //! Choose next source.
+            _log->info("[ForwardChainer] setting next source");
+
+            _fcmem.set_source(choose_next_source(_fcmem));
+
+            // Choose matching rules whose input matches with the source.
+            vector<Rule*> matched_rules = choose_rules(_fcmem);
```

Comment:
@eddiemonroe I could be wrong, but I believe the point is that for the application of the forward chainer to RelEx2Logic we want a different behavior than for the application of the forward chainer to PLN.  For R2L, what we want is to iterate through all the R2L rules, one after the other, applying each one to the {Atoms that came out of NL comprehension and are being processed via the R2L rules}.   For this use case, choosing rules via tournament selection is not appropriate and slows things down.   Recall that the R2L rules are similar to (though simpler than, and more abstract than) the RelEx2Frame rules you built a long time ago... your RelEx2Frame rule engine didn't use tournament selection but simply iterated through all the rules, applying them to the RelEx output, right?

I don't this @misgeatgit aims to replace tournament selection with "iterating through all the rules" for the PLN use-case, only for the R2L use-case....  The forward chainer should provide both options...

We discussed this extensively in the office  in HK yesterday, though the conversation was not captured and recorded online...